### PR TITLE
This prevents errors in attribute_cookie_data from rolling back all basket-related transactions.

### PR DESCRIPTION
**Pull request by schenedx**
@edx/ecommerce Please review
This is to fix the issue in ECOM-6366 for the long term

### CODE REVIEW

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 1]
```diff
@@ -120,12 +122,19 @@ def test_prepare_basket_calls_attribution_method(self):
             mock_attr_method.assert_called_with(basket, self.request)
 
     def test_attribute_cookie_data_affiliate_cookie_lifecycle(self):
-        """ Verify a basket is returned and referral captured. """
-        affiliate_id = 'test_affiliate'
-        self.request.COOKIES['affiliate_id'] = affiliate_id
+        """ Verify a basket is returned and referral captured if there are cookie info """
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: "if there is no cookie info"

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 2]
```diff
@@ -120,12 +122,19 @@ def test_prepare_basket_calls_attribution_method(self):
...
+
+        # If there are no cookie info, no referral captured
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: "If there is no cookie info, verify no referral is created."

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 3]
```diff
@@ -120,12 +122,19 @@ def test_prepare_basket_calls_attribution_method(self):
...
         basket = BasketFactory(owner=self.request.user, site=self.request.site)
         attribute_cookie_data(basket, self.request)
+        # Make sure no referral is created
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: This comment adds no value.

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 4]
```diff
@@ -120,12 +122,19 @@ def test_prepare_basket_calls_attribution_method(self):
...
+        with self.assertRaises(Referral.DoesNotExist):
+            Referral.objects.get(basket=basket)
 
+        # If there are cookie info, the referral is captured
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: "If there is cookie info, verify a referral is created."

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 5]
```diff
@@ -273,3 +282,56 @@ def test_attribute_cookie_data_multiple_cookies(self):
...
+        }
+
+        affiliate_id = 'affiliate'
+
+        self.request.COOKIES['test.edx.utm'] = json.dumps(utm_cookie)
+        self.request.COOKIES['affiliate_id'] = affiliate_id
+
+    def test_attribution_atomic_transation(self):
+        """
+            Verify db error raised from attributing cookie to basket
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: Nitpick: let's try to keep our docstrings consistent, please. We don't typically indent like this.

``` py
"""
Verify that an IntegrityError raised while creating a referral
does not prevent a basket from being created.
"""
```

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 6]
```diff
@@ -273,3 +282,56 @@ def test_attribute_cookie_data_multiple_cookies(self):
...
+        self._setup_request_cookie()
+        product = ProductFactory()
+        existing_basket = Basket.get_basket(self.request.user, self.request.site)
+        existing_referral = Referral(basket=existing_basket, site=self.request.site)
+        # Let's save an existing referral object to force the duplication happen in database
+        existing_referral.save()
+
+        with transaction.atomic():
+            with mock.patch('ecommerce.extensions.basket.utils._referral_from_basket_site') as mock_get_referral:
+                # Mock to return a duplicated referral object, so when saved, a DB integrity error is raised
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py:326]: Please add a comment about how mocking an exception as a side effect doesn't trigger a rollback. That's valuable knowledge we don't want to lose.

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 7]
```diff
@@ -273,3 +282,56 @@ def test_attribute_cookie_data_multiple_cookies(self):
         # test referral record is deleted when no cookies are set
         with self.assertRaises(Referral.DoesNotExist):
             Referral.objects.get(basket_id=basket.id)
+
+
+class BasketUtilsTranactionTests(UserMixin, TransactionTestCase):
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: BasketUtilsTransactionTests

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 8]
```diff
@@ -273,3 +282,56 @@ def test_attribute_cookie_data_multiple_cookies(self):
...
+            'utm_content': utm_content,
+            'created_at': utm_created_at,
+        }
+
+        affiliate_id = 'affiliate'
+
+        self.request.COOKIES['test.edx.utm'] = json.dumps(utm_cookie)
+        self.request.COOKIES['affiliate_id'] = affiliate_id
+
+    def test_attribution_atomic_transation(self):
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: test_attribution_atomic_transaction

File: ecommerce/extensions/basket/utils.py [hunk 9]
```diff
@@ -91,18 +92,20 @@ def get_basket_switch_data(product):
 
 def attribute_cookie_data(basket, request):
     try:
-        referral = _referral_from_basket_site(basket, request.site)
+        with transaction.atomic():
+            # Required to prevent rolling back other atomic commits if exception is found
```
Comment (rlucioni) [inline ecommerce/extensions/basket/utils.py]: You can add a little more detail here:

``` py
# If an exception is raised below, this nested atomic block prevents the
# outer transaction created by ATOMIC_REQUESTS from being rolled back.
```

**Review by rlucioni — CHANGES_REQUESTED** [review]

File: requirements/test.txt [hunk 10]
```diff
@@ -2,7 +2,7 @@
 -r base.txt
 
 bok-choy==0.4.7
-coverage==4.0.2
+coverage==4.2
```
Comment (MatthewPiatetsky) [inline requirements/test.txt:5]: is there a reason we're not specifying the patch version here?

Comment (rlucioni) [reply requirements/test.txt:5]: @MatthewPiatetsky the coverage package doesn't specify one: https://pypi.python.org/pypi/coverage.

File: ecommerce/extensions/basket/utils.py [hunk 11]
```diff
@@ -91,18 +92,20 @@ def get_basket_switch_data(product):
...
+            # Save the record if any attribution attributes are set on it.
+            if any([getattr(referral, attribute) for attribute in Referral.ATTRIBUTION_ATTRIBUTES]):
+                referral.save()
+            # Clean up the record if no attribution attributes are set and it exists in the DB.
+            elif referral.pk:
+                referral.delete()
+            # Otherwise we can ignore the instantiated but unsaved referral
 
     # Don't let attribution errors prevent users from creating baskets
     except:  # pylint: disable=broad-except, bare-except
```
Comment (MatthewPiatetsky) [inline ecommerce/extensions/basket/utils.py:112]: Do we want to differentiate an IntegrityError due to the atomic condition vs other unrelated errors?

Comment (rlucioni) [reply ecommerce/extensions/basket/utils.py:112]: Any exceptions raised here will trigger a rollback of the inner block.

File: ecommerce/extensions/basket/tests/test_utils.py (hunk 1 above)
Comment (schenedx) [reply ecommerce/extensions/basket/tests/test_utils.py]: @rlucioni should be `if there is cookie info`. We should not have referral captured if there is no cookie info

**Comment by schenedx** [comment]
@rlucioni Could you please review again?

File: ecommerce/extensions/basket/utils.py (hunk 11 above)
Comment (bderusha) [reply ecommerce/extensions/basket/utils.py:112]: @MatthewPiatetsky I don't think we do.  The point is to log, but not raise any issues that happen during attribution

**Review by bderusha — APPROVED** [review]
👍

File: ecommerce/extensions/basket/tests/test_utils.py [hunk 12]
```diff
@@ -273,3 +281,60 @@ def test_attribute_cookie_data_multiple_cookies(self):
...
+        # Let's save an existing referral object to force the duplication happen in database
+        existing_referral.save()
+
+        with transaction.atomic():
+            with mock.patch('ecommerce.extensions.basket.utils._referral_from_basket_site') as mock_get_referral:
+                # Mock to return a duplicated referral object, so when saved, a DB integrity error is raised
+                # Mocking to raise random exception will not roll back the transaction
+                # Django db would not be aware of the mocked exception,
+                # since we actually would handle the exception in the attribute_cookie_data method.
+                # This would result in the transaction being committed
```
Comment (rlucioni) [inline ecommerce/extensions/basket/tests/test_utils.py]: Thanks for adding this. However, are we sure that what you've written is entirely true? We don't want to mislead future maintainers with false assumptions. We know that using a mock side effect to raise an exception isn't enough to trigger a rollback, and we know that forcing a genuine IntegrityError is. Unless we know why that is, I wouldn't go any further here.

**Review by rlucioni — COMMENTED** [review]
@schenedx looks good, just one last small thing.

**Comment by schenedx** [comment]
@rlucioni Took in your last feedback. Can you review and comment?

**Review by rlucioni — APPROVED** [review]

**Pull request merged**